src/hotspot/share/runtime/objectMonitor.cpp

```diff
@@ -402,29 +402,34 @@ bool ObjectMonitor::enter(JavaThread* current) {
     }
 
     OSThreadContendState osts(current->osthread());
-    ThreadBlockInVM tbivm(current);
 
-    // TODO-FIXME: change the following for(;;) loop to straight-line code.
-    for (;;) {
-      current->set_suspend_equivalent();
-      // cleared by handle_special_suspend_equivalent_condition()
-      // or java_suspend_self()
+    assert(current->thread_state() == _thread_in_vm, "invariant");
 
+    current->frame_anchor()->make_walkable(current);
+    OrderAccess::storestore();
+    for (;;) {
+      current->set_thread_state(_thread_blocked);
```

Comment:
So IIUC because ThreadblockInVM  is now responsible for handling thread suspension checks, but those checks need to be done at a lower level, we can no longer use ThreadBlockInVM in some places. That both undermines the value of ThreadBlockInVM and make this manual management code much more complex.
I agree with Richard that a TBIVM that took an unlock function would make this kind of code much clearer.

Reply:
The problem is the we lock the OM from blocked state and then regret it.
Since we should not execute 'bytecode' during safe state but this implementation does that we end up in this mess.
If this implementation did the correct thing we could use ThreadblockInVM straight forward.
As I suggested if we are only blocked around parker it fixes it.

The reason for grabbing the lock is that we cannot call exit on the OM which wakes next thread.
Only the owner may call exit, this makes succession more complicate because you might wake a suspended thread.
Now the suspended thread needs to wake another thread without owning the lock.